From 86e6a395f8637ca2bed6ab1cb7cc29c9e50684ac Mon Sep 17 00:00:00 2001
From: Stuart Menefy <stuart.menefy@codasip.com>
Date: Wed, 17 Jan 2024 14:38:16 +0000
Subject: [PATCH 09/17] target/cpu: Update A730 extensions

Take the extension list from the A730 architecture manual (which is
almost the same as the list in the RVA22 profile list), and where
where an extension is supported by QEMU enable it.

Note that currently the Hobgoblin platform instantiates a A730 without
a PMP, although it could have one in the future.

Signed-off-by: Stuart Menefy <stuart.menefy@codasip.com>
---
 target/riscv/cpu.c | 45 ++++++++++++++++++++++++++++++++++++++-------
 1 file changed, 38 insertions(+), 7 deletions(-)

diff --git a/target/riscv/cpu.c b/target/riscv/cpu.c
index 71f83a7d5d..4df5241a96 100644
--- a/target/riscv/cpu.c
+++ b/target/riscv/cpu.c
@@ -511,15 +511,46 @@ static void rv64_codasip_a730_cpu_init(Object *obj)
     RISCVCPU *cpu = RISCV_CPU(obj);
 
     set_misa(env, MXL_RV64, RVI | RVM | RVA | RVF | RVD | RVC | RVS | RVU);
-    env->priv_ver = PRIV_VERSION_1_12_0;
+    cpu->cfg.mmu = true;
+    cpu->cfg.pmp = false;
+
+    cpu->cfg.ext_icsr = true;           // Zicsr
+    // Zicntr
+    cpu->cfg.pmu_num = 4;               // Zihpm
+    // Ziccif, Ziccrse, Ziccamoa
+    // Za64rs
+    // Zicclsm
+    cpu->cfg.ext_zihintpause = true;    // Zihintpause
+    cpu->cfg.ext_zba = true;            // Zba
+    cpu->cfg.ext_zbb = true;            // Zbb
+    cpu->cfg.ext_zbs = true;            // Zbs
+    cpu->cfg.ext_zcb = true;            // Zcb
+    // Zic64b
+    cpu->cfg.ext_icbom = true;          // Zicbom
+    cpu->cfg.cbom_blocksize = 64;
+    // Zicbop
+    cpu->cfg.ext_icboz = true;          // Zicboz
+    cpu->cfg.cboz_blocksize = 64;
+    cpu->cfg.ext_zfhmin = true;         // Zfhmin
+    cpu->cfg.ext_zkt = true;            // Zkt
+    cpu->cfg.ext_ifencei = true;        // Zifencei
+    // Zam
+    env->priv_ver = PRIV_VERSION_1_12_0; // Ss1p12
+    // Svbare
 #ifndef CONFIG_USER_ONLY
-    set_satp_mode_max_supported(cpu, VM_1_10_SV39);
+    set_satp_mode_max_supported(cpu, VM_1_10_SV39); // Sv39
 #endif
-
-    cpu->cfg.ext_ifencei = true;
-    cpu->cfg.ext_icsr = true;
-    cpu->cfg.mmu = true;
-    cpu->cfg.pmp = true;
+    // Svade
+    // Sscounterenw
+    // Ssccptr
+    // Sstvecd
+    // Sstvala
+    cpu->cfg.ext_svpbmt = true;         // Svpbmt
+    cpu->cfg.ext_svinval = true;        // Svinval
+    // Sm1p12
+    // Ssu64xl
+    // Sdext
+    // Sdtrig
 
     cpu->cfg.mvendorid = CODASIP_MVENDORID;
     cpu->cfg.marchid = CODASIP_MARCHID;
-- 
2.39.3

